CODING_RULES.md

```diff
@@ -11,90 +11,156 @@ to be completed
-
-  - raise the exception as a fedbiomed exception
-
-
-- caller/intermediate side: trap the exceptions:
-
-  - separate the FedbiomedError and other exceptions
-
-```
-try:
-
-    something()
-
-except SysError as e:
-    logger.XXX()
-    raise OneOfFedbiomedError()
-```
-
-  - example of top level function (eg: Experiment())
-
-```
-try:
-
-    something()
-
-except FedbiomedError as e:
-    etc...
-
-except Exception as e:   <=== objective: minimize the number of type we arrive here !
-    # place to do a backtrace
-    # extra message to the end user to post this backtrace to the support team
-    etc...
-```
-
-  - except of the top level program, it is **forbidden** to trap all exceptions (with ```except:``` or ```except Exception```)
-
-
-- the **try:** block is as small as possible
-
-- force to read the documentation
-
+### exceptions handling
+
+- on the node: in general, node should not stop because of exceptions that occur while executing requests received from researcher
+
+- on the researcher: general behaviour is to propagate the exceptions to the top level layer, where they are transformed to a friendlier output. Researcher displays this output and stops.
+
+- when a class raises an exception, it raises a FedbiomedSomethingError: use exceptions defined in **fedbiomed.common.exceptions** :
+
+  Do:
+  ```
+  raise FedbiomedSometypeError()
+  ```
+
+  Don't:
+  ```
+  raise NameError()
+  ```
+
+- optionally, if more specificity is wanted, a class can catch a python (non-Fed-Biomed) exception and re-raise it as a FedbiomedError
+
+  Optionally do:
+  ```
+  try:
+      something()
+  except SysError as e:
+      raise OneOfFedbiomedError()
+  ```
+
+- a class generally shouldn't catch a FedbiomedError and re-raise (a FedbiomedError):
+
+  Don't:
+  ```
+  try:
+    somecode()
+  except FedbiomedSometypeError;
+    raise FedbiomedOnetypeError()
+  ```
+
+- when catching exceptions
+
+  - try to be specific about the exception if easy/meaningful:
+
+    Ideally:
+    ```
+    try:
+      mycode()
+    except ValueError as e:
+      ...
+    ```
+
+    If needed:
+    ```
+    try:
+      mycode()
+    except Exception as e:
+      ...
+    ```
+
+  - don't use the `except:` clause
+
+    Don't
+    ```
+    try:
+      mycode()
+    except:
+      ...
+    ```
+
+  - should separate FedbiomedError and other exceptions
+
+    Do:
+    ```
+    try:
+      mycode()
+    except FedbiomedSomeError as e:
+      ...
+    except FedbiomedError as e:
+      ...
+    except Exception as e:
+      ...
+    ```
+
+    Don't (in case where the Exception can be a FedbiomedError)
+    ```
+    try:
+      some_code()
+    except Exception as e:
+      ...
+    ```
+
+- in general, a class shouldn't log in logger when raising or re-raising an exception. The class should log when catching and not re-raising a FedbiomedError exception. The class can log when catching a and not re-raising a python exception.
+
+  Do:
```

Comment:
I would complete the above  with something like: 

" The idea is to not trigger exception so `Node` process is not killed"

Reply:
Thank you @ybouilla 

I think this exists in this paragraph (at the beginning of the exception handling rules, because this is not specific to log/logger):

> on the node: in general, node should not stop because of exceptions that occur while executing requests received > from researcher. Top level layer code should catch and handle the exceptions, and can send an error message to 
> the researcher (but without full exception message to avoid leaking information).